posydon/unit_tests/popsyn/test_norm_pop.py

```diff
@@ -0,0 +1,839 @@
+        assert np.allclose(result, np.ones_like(m_test))
+    
+    
+    def test_valid_imf_returns_dummy_pdf(self, monkeypatch):
+        # Use monkeypatch to add DummyIMF to IMFs module
+        monkeypatch.setattr(IMFs, 'DummyIMF', DummyIMF, raising=False)
+        
+        kwargs = {
+            'primary_mass_scheme': 'DummyIMF',
+            'primary_mass_min': 1,
+            'primary_mass_max': 100
+        }
+        pdf_func = norm_pop.get_IMF_pdf(kwargs)
+        m_test = np.array([1, 10, 50])
+        result = pdf_func(m_test)
+        # Expect DummyIMF.pdf to return 0.5 regardless of m1
+        expected = 0.5
+        assert np.allclose(result, expected)
+
+    def test_imf_with_scheme_kwargs(self, monkeypatch):
+        # Create a test IMF class that uses scheme kwargs
+        class SchemeKwargsIMF:
+            def __init__(self, m_min, m_max, custom_param=None):
+                self.m_min = m_min
+                self.m_max = m_max
+                self.custom_param = custom_param
+            def pdf(self, m1):
+                # Return custom_param if set, otherwise 1
+                return self.custom_param if self.custom_param is not None else 1
+        
+        # Use monkeypatch to add SchemeKwargsIMF to IMFs module
+        monkeypatch.setattr(IMFs,
+                            'SchemeKwargsIMF',
+                            SchemeKwargsIMF,
+                            raising=False)
+        
+        # Create kwargs with scheme-specific parameters
+        kwargs = {
+            'primary_mass_scheme': 'SchemeKwargsIMF',
+            'primary_mass_min': 1,
+            'primary_mass_max': 100,
+            'SchemeKwargsIMF': {'custom_param': 3.5}  # Scheme-specific kwargs
+        }
+        
+        # Get the PDF function
+        pdf_func = norm_pop.get_IMF_pdf(kwargs)
+        
+        # Test that the custom parameter was used
+        m_test = np.array([1, 10, 50])
+        result = pdf_func(m_test)
+        expected = kwargs['SchemeKwargsIMF']['custom_param']  # The value of custom_param
+        assert np.allclose(result, expected)
+        
+
+
+class TestGetMassRatioPdf:
+    def test_flat_mass_ratio_pdf(self):
+        # kwargs without q_min and q_max to trigger flat_mass_ratio branch
+        kwargs = {
+            'secondary_mass_scheme': 'flat_mass_ratio',
+            'secondary_mass_min': 5,
+            'secondary_mass_max': 50
+        }
+        q_pdf = norm_pop.get_mass_ratio_pdf(kwargs)
+        # For m1 = 10, minimum = max(5/10, 0.05) = 0.5 and
+        # maximum = min(50/10, 1) = 1, so pdf = 1/(1-0.5) = 2.
+        m1_val = 10
+        # Test with q in range
+        result = q_pdf(0.75, m1_val)
+        assert np.allclose(result, 2)
+        # Test with q below the minimum
+        result = q_pdf(0.3, m1_val)
+        assert np.all(result == 0)
+        # Test with q above the maximum
+        result = q_pdf(1.1, m1_val)
+        assert np.all(result == 0)
```

Comment:
As this test specifies secondary masses, it's easier to see the ranges when passing m2/m1.
```suggestion
        # Test with q in range
        result = q_pdf(7.5/m1_val, m1_val)
        assert np.allclose(result, 2)
        # Test with q below the minimum
        result = q_pdf(3.0/m1_val, m1_val)
        assert np.all(result == 0)
        # Test with q above the maximum
        result = q_pdf(11.0/m1_val, m1_val)
        assert np.all(result == 0)
```